File: src/test/java/com/truveta/opentoken/unit/tokentransformer1/EncryptTokenTransformerTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.unit.tokentransformer;

import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.util.Base64;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.truveta.opentoken.tokentransformer.EncryptTokenTransformer;

public class EncryptTokenTransformerTest {
    private EncryptTokenTransformer transformer;
    private static final String VALID_KEY = "12345678901234567890123456789012"; // 32-byte key
    private static final String INVALID_KEY = "short-key"; // Invalid short key

    @BeforeEach
    public void setUp() throws Exception {
        transformer = new EncryptTokenTransformer(VALID_KEY);
    }

    @Test
    public void testConstructor_ValidKey_Success() throws Exception {
        EncryptTokenTransformer validTransformer = new EncryptTokenTransformer(VALID_KEY);
        Assertions.assertNotNull(validTransformer);
    }

    @Test
    public void testConstructor_InvalidKeyLength_ThrowsIllegalArgumentException() {
        Exception exception = Assertions.assertThrows(InvalidKeyException.class, () -> {
            new EncryptTokenTransformer(INVALID_KEY); // Key is too short
        });
        Assertions.assertEquals("Key must be 32 characters long", exception.getMessage());
    }

    @Test
    public void testTransform_ValidToken_ReturnsEncryptedToken() throws Exception {
        String token = "mySecretToken";
        String encryptedToken = transformer.transform(token);

        // Ensure the encrypted token is not null or empty
        Assertions.assertNotNull(encryptedToken);
        Assertions.assertFalse(encryptedToken.isEmpty());

        // check if the token is base64-encoded by decoding it
        byte[] decodedBytes = Base64.getDecoder().decode(encryptedToken);
        Assertions.assertNotNull(decodedBytes);
    }

    @Test
    public void testTransform_ReversibleEncryption() throws Exception {
        // Testing if encryption followed by decryption will give back the original
        // token.
        String token = "mySecretToken";

        String encryptedToken = transformer.transform(token); // Encrypt the token

        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding"); // Decrypt the token using the same settings
        SecretKeySpec secretKey = new SecretKeySpec(VALID_KEY.getBytes(), "AES");
        IvParameterSpec iv = new IvParameterSpec(new byte[16]); // 16-byte IV (all zeroes)
        cipher.init(Cipher.DECRYPT_MODE, secretKey, iv);
        byte[] decryptedBytes = cipher.doFinal(Base64.getDecoder().decode(encryptedToken));
        String decryptedToken = new String(decryptedBytes, StandardCharsets.UTF_8);

        Assertions.assertEquals(token, decryptedToken); // Ensure the decrypted token matches the original token
    }
}

File: lib/nodejs/opentoken/test/tokentransformer/EncryptTokenTransformer.test.ts
/**
 * Copyright (c) Truveta. All rights reserved.
 */

import { EncryptTokenTransformer } from '../../src/tokentransformer/EncryptTokenTransformer';

describe('EncryptTokenTransformer', () => {
  const testKey = 'Secret-Encryption-Key-Goes-Here.'; // 32 characters

  it('should encrypt a token using AES-256-GCM', () => {
    const transformer = new EncryptTokenTransformer(testKey);
    const token = 'test-token-value';
    const encrypted = transformer.transform(token);

    // Should return a base64-encoded string
    expect(encrypted).toBeDefined();
    expect(encrypted.length).toBeGreaterThan(0);
    expect(encrypted).toMatch(/^[A-Za-z0-9+/=]+$/);
  });

  it('should produce different encrypted values each time (due to random IV)', () => {
    const transformer = new EncryptTokenTransformer(testKey);
    const token = 'same-token';
    
    const encrypted1 = transformer.transform(token);
    const encrypted2 = transformer.transform(token);

    // Different because of random IV each time
    expect(encrypted1).not.toBe(encrypted2);
  });

  it('should encrypt different tokens to different values', () => {
    const transformer = new EncryptTokenTransformer(testKey);
    
    const encrypted1 = transformer.transform('token1');
    const encrypted2 = transformer.transform('token2');

    expect(encrypted1).not.toBe(encrypted2);
  });

  it('should throw error for null or empty token', () => {
    const transformer = new EncryptTokenTransformer(testKey);
    
    expect(() => transformer.transform('')).toThrow('Token cannot be null or empty');
    expect(() => transformer.transform('  ')).toThrow('Token cannot be null or empty');
  });

  it('should throw error for invalid key length', () => {
    expect(() => new EncryptTokenTransformer('short')).toThrow('Key must be 32 characters long');
    expect(() => new EncryptTokenTransformer('way-too-long-key-that-exceeds-32-chars')).toThrow('Key must be 32 characters long');
  });

  it('should accept 32-character key', () => {
    expect(() => new EncryptTokenTransformer(testKey)).not.toThrow();
  });
});
